diff --git a/KernelSU/kernel/Makefile b/KernelSU/kernel/Makefile
index 6ff901f..9a7fb75 100644
--- a/KernelSU/kernel/Makefile
+++ b/KernelSU/kernel/Makefile
@@ -20,9 +20,19 @@ $(eval KSU_VERSION=$(shell expr 10000 + $(KSU_GIT_VERSION) + 200))
 $(info -- KernelSU version: $(KSU_VERSION))
 ccflags-y += -DKSU_VERSION=$(KSU_VERSION)
 else # If there is no .git file, the default version will be passed.
+ifdef import_KSU_GIT_VERSION
+$(info import_KSU_GIT_VERSION:\${import_KSU_GIT_VERSION})
+$(info -- "Used KSU_GIT_VERSION imported externally !")
+KSU_GIT_VERSION = \$(import_KSU_GIT_VERSION)
+# ksu_version: major * 10000 + git version + 200 for historical reasons
+$(eval KSU_VERSION=\$(shell expr 10000 + \$(KSU_GIT_VERSION) + 200))
+$(info -- set KernelSU version: \$(KSU_VERSION))
+ccflags-y += -DKSU_VERSION=$(KSU_VERSION)
+else
 $(warning "KSU_GIT_VERSION not defined! It is better to make KernelSU a git submodule!")
 ccflags-y += -DKSU_VERSION=16
 endif
+endif
 
 ifndef EXPECTED_SIZE
 EXPECTED_SIZE := 0x033b
-- 
2.41.0.windows.1

